kafka-client/src/main/java/dev/responsive/kafka/api/stores/TtlProvider.java

```diff
@@ -0,0 +1,285 @@
+      return ttl.toSeconds();
+    }
+
+    public long toMillis() {
+      return ttl.toMillis();
+    }
+
+    @Override
+    public boolean equals(final Object o) {
+      if (this == o) {
+        return true;
+      }
+      if (o == null || getClass() != o.getClass()) {
+        return false;
+      }
+
+      final TtlDuration that = (TtlDuration) o;
+
+      if (!ttl.equals(that.ttl)) {
+        return false;
+      }
+      return ttlType == that.ttlType;
+    }
+
+    @Override
+    public int hashCode() {
+      int result = ttl.hashCode();
+      result = 31 * result + ttlType.hashCode();
+      return result;
+    }
+  }
+
+  private enum TtlType {
+    DEFAULT_ONLY,
+    KEY,
+    VALUE,
+    KEY_AND_VALUE
+  }
+
+  private final TtlType ttlType;
+  private final TtlDuration defaultTtl;
+
+  // Only non-null for key/value-based ttl providers
+  private final Serde<K> keySerde;
+  private final Serde<V> valueSerde;
+
+  private final BiFunction<K, V, Optional<TtlDuration>> computeTtl;
+
+  private TtlProvider(
+      final TtlType ttlType,
+      final TtlDuration defaultTtl,
+      final BiFunction<K, V, Optional<TtlDuration>> computeTtl,
+      final Serde<K> keySerde,
+      final Serde<V> valueSerde
+  ) {
+    this.ttlType = ttlType;
+    this.defaultTtl = defaultTtl;
+    this.keySerde = keySerde;
+    this.valueSerde = valueSerde;
+    this.computeTtl = computeTtl;
+  }
+
+  public Serde<K> keySerde() {
+    return keySerde;
+  }
+
+  public Serde<V> valueSerde() {
+    return valueSerde;
+  }
+
+  public TtlDuration defaultTtl() {
+    return defaultTtl;
+  }
+
+  public boolean hasConstantTtl() {
+    return ttlType == TtlType.DEFAULT_ONLY;
+  }
+
+  public boolean needsValueToComputeTtl() {
+    return ttlType == TtlType.VALUE || ttlType == TtlType.KEY_AND_VALUE;
+  }
+
+  public Optional<TtlDuration> computeTtl(
+      final byte[] keyBytes,
+      final byte[] valueBytes,
+      final StateDeserializer<K, V> stateDeserializer
+  ) {
+    final K key;
+    final V value;
+    switch (ttlType) {
+      case DEFAULT_ONLY:
+        key = null; //ignored
+        value = null; //ignored
+        break;
+      case KEY:
+        key = stateDeserializer.keyFrom(keyBytes);
+        value = null; //ignored
+        break;
+      case VALUE:
+        key = null; //ignored
+        value = stateDeserializer.valueFrom(valueBytes);
+        break;
+      case KEY_AND_VALUE:
+        key = stateDeserializer.keyFrom(keyBytes);
+        value = stateDeserializer.valueFrom(valueBytes);
+        break;
+      default:
+        throw new IllegalStateException("Unrecognized ttl type: " + ttlType);
+    }
+    final Optional<TtlDuration> rowTtlOverride = computeTtl.apply(key, value);
+
+    // Just return Optional.empty if the row override is actually equal to the default
+    // to help simplify handling logic for tables
+    if (rowTtlOverride.isPresent() && rowTtlOverride.get().equals(defaultTtl)) {
+      return Optional.empty();
+    }
```

Comment:
I'd prefer we removed this optimization unless it proves necessary -- does it actually simplify things? I imagine we shouldn't be comparing the result of this to the default anywhere in the code

It works with today's implementation, but if we ever have partial updates it's risky (as in I inserted a row with a non-default TTL and then I update it to be the default TTL I should make sure that every scenario would override the old TTL)